---
title: "DMR-2-GeneIDs"
output: html_document
---

In this Rmarkdown file we will take a methylkit output file and end up with a list of genes associated with the DMRs based on location. Files being provided to you (results of other processes) are in the `data` directory.



Loading libaries 
```{r, message = FALSE}
library(dplyr)
library(reshape2)
```


# Step  1. PREPARING A BED FILE

Here we read in the methylKit output file containing DMRs between sperm and RBCs in O.mykiss

```{r}
DMRoutput <- read.csv ("../data/SpvRBC_DMR100bp_75percdiff")
knitr::kable(
  DMRoutput[1:5, ]
)
```



Filtering DMRs to return only those that are hypo-methylated (in sperm v. RBC)

```{r}
DMRoutput_hypo <- filter(DMRoutput, meth.diff < 0)
knitr::kable(
  DMRoutput_hypo[1:5, ]
)
```


Reformatting the output into a bed file (fields: chrom, chromStart,chromEnd,name*)

The name field can contain any information that you want to keep with the region, we will retain the meth.diff column. 

The third line of code in this chunk writes out the table to the `analyses` directory.


```{r}
DMRoutput_hypo.bed <- select(DMRoutput_hypo,chr,start,end,meth.diff)  
knitr::kable(
  DMRoutput_hypo.bed[1:5, ]
)
write.table (DMRoutput_hypo.bed,"../analyses/DMRs_hypo.bed",col.names = F, sep = '\t', row.names = F,quote = F)
```




# STEP 2. USING GENOMIC ARITHMATIC TO ANNOTATE REGIONS OF INTEREST

The first thing you will need to do is revise the chunk below to include the locations of bedtools on your computer.

```{bash}
bedtools=/Applications/bioinfo/bedtools2/bin/bedtools
```


First we will sort the file created in prior Step.

```{bash}
$bedtools \
sort \
-i ../analyses/DMRs_hypo.bed \
> ../analyses/DMRs_hypo.sorted.bed
```


Then find closest genes with -D arguement reporting distance from genes

```{bash}
/Applications/bioinfo/bedtools2/bin/bedtools \
closest \
-a ../analyses/DMRs_hypo.sorted.bed \
-b ../data/Omy_mRNA_subset.gff \
-D b \
> ../analyses/closetgene.txt

head ../analyses/closetgene.txt
```




# STEP 3. JOINING TABLES AND FILTERING BASED ON DISTANCE AND QUALIY OF ANNOTATION

Reading in table from Step 2 into R.

```{r}
DMRwgene <- read.table("../analyses/closetgene.txt", sep = '\t')

knitr::kable(
  DMRwgene[1:2, ]
)
```


Reading in blast output

```{r}
genes_annot <- read.table("../data/Omy_BLAST_UniProt.tab", header=T, sep="\t", stringsAsFactors=F, fill=T, quote="", na.strings="", comment.char="")


knitr::kable(
  genes_annot[1:2, ]
)
```



Performing a left join to join the output of bedtools (DMRwgene) to the blast output (genes_annot)

```{r}
geneID_IPA <- left_join(DMRwgene,genes_annot, by = c("V13" = "Sequence"))

```




```{r}
head(geneID_IPA)

```

Now we will filter the annotations to, 1) genes within 10kb of DMR, 2) and e-value of < 1e-10



```{r}
annotations_filterA<-filter(geneID_IPA, V14 >= -10000)
nrow(annotations_filterA)
annotations_filterB<-filter(annotations_filterA, V14 <= 10000)
nrow(annotations_filterB)
annotations_filterC<-filter(annotations_filterB, E.Value <= 1e-10)
nrow(annotations_filterC)
```

Extracting the 'UNIPROT_ACCESSION' and removing any duplicates for use in enrichment analysis 

Splitting column 'Hit.Name' by "|"

```{r}
newColNames <- c("DB", "UNIPROT_ACCESSTION","UNIPROT_ID")
newCols <- colsplit(annotations_filterC$Hit.Name, "\\|", newColNames)
head(newCols)
annotations_filterC_UNIPROT <- cbind(annotations_filterC, newCols)

head(annotations_filterC_UNIPROT)
```


Saving files 

```{r}
write.table(annotations_filterC_UNIPROT,"../analyses/DMR_fullannotations.txt",sep = "\t",quote = F,row.names = F)


write.table(annotations_filterC_UNIPROT$UNIPROT_ID, "../analyses/DMR_UNIPROT_ID.txt", sep = "\t", quote = F, row.names = F, col.names = F)
```








